src/Symfony/Component/Clock/HrClock.php

```diff
@@ -0,0 +1,69 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Clock;
+
+/**
+ * @author Nicolas Grekas <p@tchwork.com>
+ */
+class HrClock implements TimeInterface
```

Comment:
I'd add the ClockInterface and a possible SleepInterface as well here to separate the different concerns better.

* the ClockInterface to get the current Point in Time (which can then be mocked to represent any point in time)
* the TimeInterface for a convenient way of getting the time including microseconds (which could perhaps also contain a `timeInt` method to get the timestamp as integer),
* the SleepInterface for a convenient way to control the flow of the application. As we have discussed regarding PSR-20 `sleep` has nothing to do with getting a point of time but with controlling the flow of the application.

Reply:
I don't think splitting in many interfaces would provide any benefits: all those methods are tightly related and fit a single abstraction to me. Having a single `TimeInterface` also provides better DX, with no drawbacks IMHO.

> which could perhaps also contain a timeInt method to get the timestamp as integer

Good idea, PR updated.

Reply:
> I don't think splitting in many interfaces would provide any benefits: all those methods are tightly related and fit a single abstraction to me. Having a single `TimeInterface` also provides better DX, with no drawbacks IMHO.

To me they are three completely separate use-cases:

* Getting a dedicated Point in the dimension of time. (`now`)
* getting a number of seconds after a specific point in time (`timeArray` and `timeFloat`)
* stopping the execution of the script for a dedicated amount of time.

Especially the last one seems to be something unrelated to timekeeping in terms of *the interface*. It might absolutely make sense to have them within *the implementation*  IMO.

But that feels like a debate of principles and in the end I'm not the one taking that decision for symfony (luckily :wink:).

Reply:
> Especially the last one seems to be something unrelated to timekeeping in terms of the interface. It might absolutely make sense to have them within the implementation IMO.

I don't have the same view on that topic: to me, sleep() is like a setter and the other methods are sort of getters. Segregating setters from getters doesn't make sense when they work on the same state.

About segregating `now` and `time*` I'd prefer not because it makes using the abstraction more complex for little to no practical benefits. Providing them at the abstraction level on the other end ensures that all those methods are always provided by implementations, enforcing the synchronization of the returned time by all methods.

Reply:
The method `sleep` could be decorated with a stopwatch so that sleeping time can be visualized in the profiler. I would say that having less methods in the interface is better for decoration.